fplll/pruner/pruner.h

```diff
@@ -290,37 +382,186 @@ template <class FT> class Pruner
+     target unciton is:
+
+     single_enum_cost(pr) * trials + preproc_cost * (trials - 1.0);
+  */
+  void optimize_coefficients_cost(/*io*/ vector<double> &pr);
+
+  /**
+     @brief main interface to optimize the single enumeraiton time fixing succ. prob.
+
+     Main interface to optimize the single enumeration time with the constraint
+     that the succ. prob (or expected solutions) is fixed.
+  */
+  void optimize_coefficients_prob(/*io*/ vector<double> &pr);
+
+  /**
+     @brief auxiliary function in optimizing the single enumeraiton time
+     fixing the succ. prob.
+
+     Auxiliary function to optimize the single enumeration time with the
+     constraint that the succ. prob (or expected solutions) is fixed. It
+     is used if the given targeted probaility is larger than the one
+     computed using pruning parameters. Then one increases the succ.
+     probability by increasing the pruning parameters. Note since the
+     radius is fixed as input, it may not be possible to achieve the
+     given targeted probability even if all coefficients are (1, ..., 1).
+     In such case, it will try to increase the succ. prob as much as
+     it can and possibly return (1, ..., 1).
+  */
+  void optimize_coefficients_prob_incr(/*io*/ vector<double> &pr);
+
+  /**
+     @brief auxiliary function in optimizing the single enumeraiton time
+     fixing the succ. prob.
+
+     Auxiliary function to optimize the single enumeration time with the
+     constraint that the succ. prob (or expected solutions) is fixed. It
+     is used if the given targeted probaility is smaller than the one
+     computed using pruning parameters. Then one decreases the succ.
+     probability by descreasing the pruning parameters.
+  */
+  void optimize_coefficients_prob_decr(/*io*/ vector<double> &pr);
+
+  /**
+     @brief auxiliary function in optimizing the single enumeraiton time fixing succ. prob.
+
+     Heuristic tuning procedure which seems to be useful. This is used to make the ratio
+     between the succ. prob and the target succ. prob are close. Depending on whether
+     the succ. prob is larger (or smaller), it will try to reduce the pruning coefficients
+     (or increase) to make succ. prob \approx the target succ. prob. Sometimes the succ.
+     prob can not be modified closer to target succ. prob due to the contraints.
+     In such case it exists.
+  */
+  void optimize_coefficients_prob_tune(/*io*/ vector<double> &pr);
 
+  /**
+     @brief Main interface to optimize pruning coefficients
+
+     Main interface to optimize pruning coefficients.
+     It will in turn invoke either of the two functions:
+      (1)   optimize_coefficients_cost() or
+      (2)   optimize_coefficients_prob()
+     depending on the input "target".
+
+     If the target is negative (e.g. -1), it calls function (1)
+     such that goal is to optimize the
+     single_enum_cost(pr) * trials + preproc_cost * (trials - 1.0).
+
```

Comment:
I'm still not a fan of overloading `target` (given to the constructor) like this. The decision of whether to optimize globally or a single enumeration could be handled by a flag e.g. `PRUNER_SINGLE` or so. Also, why break existing code by changing the previous default behaviour, i.e. global optimization is used when target < 0? In `master` the user supplies a `target > 0` and gets a global optimisation, in this PR the user gets a local optimisation. Is that really the more common target?

Reply:
Do you mean the following behavior? If flag is False, we ignore the target and do the optmize_coefficients_cost(). If flag is True, take the target into account and call the optmize_coefficients_prob(). 

I am just not sure if we should pass the PRUNER_SINGLE in the same existing flag. As the current flags 
"PRUNER_CVP  PRUNER_START_FROM_INPUT  PRUNER_GRADIENT  PRUNER_NELDER_MEAD  PRUNER_VERBOSE"
seem to be defined for the optimization "strategies"; but not for the "goals" of optimization. 

Or we could have an extra flag in the Pruner constructor which may be more clear.

Reply:
Aren't we using `target` either way? We got three cases:
(a) find shortest vector with probability `target` with minimal overall cost
(b) find `target` vectors within the given radius
(c) find shortest vector with probability `target` in a single enumeration 

If I understand correctly, you are currently switching between (a) and (c) by using positive and negative signs for `target`. Isn't then (c) a special case of (a) but where the strategy is to use only one enumeration?

Reply:
I thought I am switching between (b) and (c). Here is my understanding, there are two optimization goals:

Goal 1. Overall cost optimization:

single_enum_cost(b) * trials + preproc_cost * (trials - 1.0);
where the trials is computed either using svp_probability() or expected_solutions()

In such case, the target will be ignored since we don't fix the svp_probability() or expected_solutions(). We just want the overall cost to be small; the svp_probability() or expected_solutions()  may vary. I think this correspond to the case (a) above if using svp_probability().

Goal 2. Single enumeration cost optimization fixing probability/expectation.

single_enum_cost(b) while fixing `target' from input. 

In such case, we will try to achieve the `target' probability (or expected solutions) while reducing the single_enum_cost(). Still it may not guaranteed we can achieve this since the two inputs `target' and `radius' could be contradictory to each other. E.g. one set radius smaller than the GH and ask `target' very large. In these cases, it will try it's best to achieve the `target'. 

For either of the above 2 goals, there are two benchmarks:

One can use either svp_probability() or expected_solutions(). If one uses svp_probability(), then the input `radius' may not have any affects. If one uses expected_solutions(), the input `radius' may have some affects.

Reply:
This is done :) Now the optimization with fixed probability is triggered by the PRUNER_SINGLE with internal opt_single. This is compatible hopefully with current behavior in the master branch.